### PR TITLE
Extended allowlist

### PR DESCRIPTION
**Pull request by Martoon-00**
Add another implementation of allowlist to English auction and Marketplace contracts that
* Aims at an arbitrary number of addresses;
* Allows limiting permitted `token_id`s as well.

### CODE REVIEW

File: packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo [hunk 1]
```diff
@@ -0,0 +1,23 @@
+(* Base allowlisting module.
+ * Import it in a contract if it is planned to be extended with allowlisting
+ * extensions.
```
Comment (pasqu4le) [inline packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo:3]: I think this should better explain that only using this module won't actually add any allowlisting (on its own).

File: packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_token.mligo [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+type allowlist = (address, allowed_token_ids) big_map
+
+type allowlist_update =
+  [@layout:comb]
+  { to_remove : address set
+  ; to_add : (address, allowed_token_ids) map
+  }
+
+type allowlist_entrypoints = allowlist_update
+  // TODO try list/set
```
Comment (pasqu4le) [inline packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_token.mligo:28]: I see you went with a `set`, is this TODO here intended to check performance differences with a `list`?

File: packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+*)
+
+#if !ALLOWLIST_BASE
+#define ALLOWLIST_BASE
+
+#if !ALLOWLIST_ENABLED
```
Comment (pasqu4le) [inline packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo:9]: I have a question: couldn't you remove `ALLOWLIST_ENABLED` (I mean everywhere) in favor of `ALLOWLIST_BASE`?

Unless I'm missing something it looks like in the contract there are some checks for `#if !ALLOWLIST_ENABLED`, but AFAIU if no allowlist is enabled this module still needs to be imported, so `ALLOWLIST_BASE` will be defined.

Note: if you do so however in `#ifelse` chains I suppose the `!ALLOWLIST_ENABLED` that is now on top should be replaced by an `ALLOWLIST_BASE` at the bottom.

File: packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_token.mligo (hunk 2 above)
Comment (Martoon-00) [reply packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_token.mligo:28]: Yeah right, thanks for the spot. Eventually, I tried to replace `(address, allowed_token_ids) map ` with a list of pairs, and that increased fees (:open_mouth:), so I'll avoid `list`s at all.

File: packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo (hunk 3 above)
Comment (Martoon-00) [reply packages/minter-contracts/ligo/fa2_modules/fa2_allowlist/allowlist_base.mligo:9]: `ALLOWLIST_ENABLED` in the first place serves to notify `allowlist_base.mligo` that some allowlisting implementation has been included and providing primitives for empty allowlisting is not necessary.

Though, from other modules' point of view, maybe the existence of `ALLOWLIST_ENABLED` complicates things. Do you think it's worth rewriting via `ALLOWLIST_BASE`?

**Comment by Martoon-00** [comment]
Hmm, it looks like now Taquito tests fail because it does not support `never` type :confused: 

There is an issue in Taquito https://github.com/ecadlabs/taquito/issues/774 and a PR was made literally today, maybe it's worth waiting for 2-3 days more, in hope they merge it soon?

Another option is to add `Update_allowlist` entrypoint only when non-empty allowlisting is enabled, but this won't fit well into types anymore, meaning that for Cleveland tests we will have to define variations of `*_entrypoints` type w/ and w/o `Update_allowlist` entrypoint too and that's an annoying amount of boilerplate.

**Review by michaeljklein — APPROVED** [review]
LGTM

**Pull request merged**